runtime/pack/dist/opt/nvim.undotree/lua/undotree.lua

```diff
@@ -0,0 +1,406 @@
+      if type(v) == 'table' then
+        for i, j in ipairs(v) do
+          if j < minseq then
+            minseq = j
+            index = k
+            node_index = i
+          end
+        end
+      elseif v < minseq then
+        assert(type(v) == 'number')
+        minseq = v
+        index = k
+      end
+    end
+
+    local node = nodes[index]
+
+    --- @param kind 'node'|'remove'|'branch'|'nochange_remove'
+    local function add_graph_line(kind)
+      table.insert(graph_lines, { kind = kind, index = index, node_count = #nodes, node = node })
+    end
+
+    if type(node) == 'number' then
+      add_graph_line('node')
+
+      local child = tree[node].child
+      if #child == 0 then
+        if index ~= #nodes then
+          add_graph_line('remove')
+        else
+          add_graph_line('nochange_remove')
+        end
+
+        table.remove(nodes, index)
+      elseif #child == 1 then
+        nodes[index] = child[1]
+      else
+        nodes[index] = child
+      end
+    else
+      assert(type(node) == 'table')
+
+      add_graph_line('branch')
+
+      table.remove(nodes, index)
+      if #node == 2 then
+        table.insert(nodes, index, math.min(unpack(node)))
+        table.insert(nodes, index, math.max(unpack(node)))
+      elseif #node > 2 then
+        table.insert(nodes, index, node[node_index])
+        table.insert(nodes, index, node)
+        table.remove(node, node_index)
+      end
+    end
+  end
+
+  for k, v in ipairs(graph_lines) do
+    if v.kind == 'remove' and (graph_lines[k + 1] or {}).kind == 'branch' then
+      v.kind = 'remove+branch'
+      v.index2 = graph_lines[k + 1].index
+      table.remove(graph_lines, k + 1)
+    end
+  end
+
+  return graph_lines
+end
+
+--- @param time integer
+--- @return string
+local function undo_fmt_time(time)
+  if time == -1 then
+    return 'origin'
+  end
+
+  local diff = os.time() - time
+
+  if diff >= 100 then
+    if diff < (60 * 60 * 12) then
+      return os.date('%H:%M:%S', time) --[[@as string]]
+    else
+      return os.date('%Y/%m/%d %H:%M:%S', time) --[[@as string]]
+    end
+  else
+    return ('%d second%s ago'):format(diff, diff == 1 and '' or 's')
+  end
+end
+
+--- @param tree vim.undotree.tree
+--- @param graph_lines vim.undotree.graph_line[]
+--- @param buf integer
+--- @param meta {[integer]:integer}
+--- @param find_seq? integer
+--- @return integer?
+local function buf_apply_graph_lines(tree, graph_lines, buf, meta, find_seq)
+  -- As in io-buffer, not vim-buffer
+  local line_buffer = {}
+  local extmark_buffer = {}
+
+  --- @type integer?
+  local found_seq
+
+  for k, v in ipairs(graph_lines) do
+    local is_last = k == #graph_lines
+
+    --- @type string?
+    local line
+    if v.kind == 'node' then
+      line = ('| '):rep(v.index - 1)
+        .. '*'
+        .. (' |'):rep(v.node_count - v.index)
+        .. '    '
+        .. v.node
+        .. '    ('
+        .. undo_fmt_time(tree[v.node].time)
+        .. ')'
```

Comment:
Just a note on lua performance: all string are interned (hashed), so each intermediate concatenation step requires a rehashing of the new string. Just this block would require eight different strings to be hashed. It's not something noticeable with a small tree, but as it gets longer it's becomes an issue. My recommendation is to use `table.concat({"a", "b", "c"}, "")` instead, as this only results in a single new string..